### DISCUSSION TITLE
How can I force a plugin to load earlier? (`rocks-config`)

### DISCUSSION THREAD

**Discussion by jumpyjacko** [lumen-oss/rocks.nvim#265] [Q&A] [answered]
I have my colourscheme working based on the `mini.base16` package and my statusline is the `heirline.nvim` package. I use the `rocks-config` helper for the configuration. I saw in the bottom of the readme that it says:
>If you need to source a plugin's scripts eagerly (for example, to load colorschemes), you can set the plugin's opt = true in rocks.toml, and then load it with vim.cmd.Rocks({"packadd", "<rock-name>"}) or require("rocks").packadd("<rock-name").

So I added `opt = true` to the `mini.base16` line in `rocks.toml` and then put in my `init.lua`, `require("rocks").packadd("mini.base16")`. Sometimes it loads before `heirline.nvim` and sometimes it still loads after `heirline.nvim` which breaks the colours on it (same behaviour as if I didn't try to eagerly load `mini.base16`).

Is `init.lua` the correct place to put the `packadd` command? Is there a better way to define load order for specific plugins?

**Comment by mrcjkb** [comment, marked as answer]
## Update

This issue has been solved since version 2.25.0.

rocks.nvim now maintains symlinks in a `site/pack/opt` directory, which makes installed plugins available on the runtimepath by the time Neovim sources init.lua, while handing over control of sourcing plugins to rocks.nvim.

Plugins with `opt = true` can now be sourced with Neovim's builtin `:packadd`, making `:Rocks packadd` and `require("rocks").packadd` obsolete.

## Original response

Hey :wave: 

This seems to me like an issue with the `heirline.nvim` plugin.
Its API allows the `colors` option to be a function, which would be useful if it would evaluate it lazily (i.e. when the colors are needed).
But that's not the case. As soon as you call `setup`, it evaluates the `colors` function.

A proper solutionto this would be to fix this in heirline.nvim, by changing [`loaded_colors`](
https://github.com/rebelot/heirline.nvim/blob/03cff30d7e7d3ba6fdc00925f015822f79cef908/lua/heirline/highlights.lua#L8) to be a function and evaluating it as needed[^1] .

[^1]: This solution would fix the issue, regardless of the plugin manager you are using (I might PR that fix later, if I find some time).

For now, you can work around this by calling `mini.base16`'s `setup` function in your `init.lua`, but this will cause an error if the plugin isn't installed[^2].

[^2]: A downside of the [setup antipattern](https://mrcjkb.dev/posts/2023-08-22-setup.html) :disappointed: 

@vhyrro is planning to work on a `config.bundles` feature for rocks-config.nvim, which would let you use a single config for multiple related plugins. That would be useful to solve this issue for rocks-config.nvim.

**Reply by mrcjkb** [reply]
https://github.com/rebelot/heirline.nvim/pull/190

**Reply by mrcjkb** [reply]
@jumpyjacko it looks like heirline expects you to set up an autocommand for colorscheme events: 

https://github.com/rebelot/heirline.nvim/blob/master/cookbook.md#theming

**Reply by jumpyjacko** [reply]
Should this be put within my setup file or somewhere else? I currently have it in my `lua/plugins/heirline.lua` file near the top.

https://github.com/jumpyjacko/nvim/blob/e81c14cf9c9eee5b99be64ffb8619085090e138b/lua/plugins/heirline.lua#L26C1-L26C58

**Reply by mrcjkb** [reply]
That seems like it should work. 
You probably also want to set the colorscheme in the mini.base16 config file, after calling its setup function.

**Reply by jumpyjacko** [reply]
Yep, it seems I was actually setting up the colourscheme entirely wrong. Thanks for the help!

https://github.com/jumpyjacko/nvim/blob/0ae5176248ee35c3b58cc06df84176fac3d16e8d/colors/modified_ocean.lua#L1
https://github.com/jumpyjacko/nvim/blob/0ae5176248ee35c3b58cc06df84176fac3d16e8d/init.lua#L5

Edit: In fact, this loads the colourscheme first, right with `init.lua`, meaning this was never a problem with `rocks.nvim`. Sorry for the trouble!
```
000.005  000.005: --- NVIM STARTING ---
...
003.091  000.043  000.043: sourcing /usr/share/nvim/runtime/ftplugin.vim
003.158  000.021  000.021: sourcing /usr/share/nvim/runtime/indent.vim
003.201  000.009  000.009: sourcing /usr/share/nvim/archlinux.vim
003.203  000.022  000.013: sourcing /etc/xdg/nvim/sysinit.vim
003.812  000.581  000.581: require('options')
005.617  001.802  001.802: require('mappings')
005.692  000.016  000.016: require('vim.fs')
005.737  000.115  000.099: require('rocks-setup')
007.967  001.276  001.276: require('mini.base16')
009.087  002.419  001.143: sourcing /home/jackson/.config/nvim/colors/modified_ocean.lua
009.096  005.881  000.964: sourcing /home/jackson/.config/nvim/init.lua
...
```
